### ISSUE TITLE
Understanding the masking in beam search

### ISSUE THREAD

**Issue by legohyl** [chaitjo/learning-tsp#3]
Hi! I would like to ask a question regarding your `update_mask` function in `Beamsearch` for NAR model.

It seems that as you advance the decoding, you map away the `new_nodes` to 0. Then, you replace this value with `1e10`. Doesn't this make the log probability very large, and hence probability == 1, at the next iteration? Shouldn't it be trying to reduce these values instead, so that the beam does not advance there?

**Comment by chaitjo** [comment]
Hi Yong Liang, I'll have a look. IIRC, the beam search is performed using logits (log probabilities) and not raw probabilities.

**Comment by legohyl** [comment]
Yes they are using logits! But since logs are monotonically increasing, isn't setting it to `1e10` increasing its probability? Unless of course you are predicting inversely, meaning that a lower value is a higher probability.

**Comment by chaitjo** [comment]
https://github.com/chaitjo/learning-tsp/blob/a7b132aa4e532b3f23e74c240c7bfbb794ebf764/utils/nar_beam_search.py#L70

I think this is the important line -- the values in the mask are being used **multiply** with the beam. So if the probability for selecting a masked node is high, the multiplication with the mask should set it to zero or close to zero. In log probs, this boils down to multiplying a number b/w -inf and 0 (logs of 0 and 1, respectively) by a mask (inf for masked nodes). Thus, the values for the masked nodes will always be -inf through this kind of masking operation.

**Comment by chaitjo** [comment]
To be honest, now that I myself am looking at this stuff after not having worked on it for a while, it can be a bit non-trivial/confusing to understand various beam search implementations. Maybe looking at the older + more commented code (which also allowed one to work with raw probs) can aid in understanding: https://github.com/chaitjo/graph-convnet-tsp/blob/master/utils/beamsearch.py

**Comment by chaitjo** [comment]
Feel free to reopen if needed.
